Plugins: Add Subresource Integrity checks

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

- Backend: Pass SHA256 hash for module.js from MANIFEST.txt to the frontend via frontendsettings.
- Frontend: Introduce system.js integrity checks for plugins loaded with the "script" loading strategy, using the module.js hash provided by thte backend

The hash is decoded from hex to bin, then re-encoded as base64 and pre-fixed with "sha256-", as that's the format that the browser expects:

```
$ shasum -b -a 256 trafficlight.js | awk '{ print $1 }'
ddfcb449445064e6c39f0c20b15be3cb6a55837cf4781df23d02de005f436811
$ shasum -b -a 256 trafficlight.js | awk '{ print $1 }'| xxd -r -p
��IDPd�ß
         �[��jU�|�x�=�_Ch%
$ shasum -b -a 256 trafficlight.js | awk '{ print $1 }' | xxd -r -p | base64
3fy0SURQZObDnwwgsVvjy2pVg3z0eB3yPQLeAF9DaBE=
```

The hash is available in either:
- `grafanaBootData.settings.panels[*].moduleHash`
- `grafanaBootData.settings.apps[*].moduleHash`
- `grafanaBootData.settings.datasources[*].meta.moduleHash`

As well as via:
```
GET /api/plugins/:pluginId/settings
```

For example:

```
{
  "name": "Worldmap Panel",
  "type": "panel",
  "id": "grafana-worldmap-panel",
  // ...
  "moduleHash": "sha256-44p13OCeajVUOKHYleL2rdZHIjlTQ70N1jBdf87dPGs="
}
```

For unsigned plugins or in case of errors, the hash will not be passed to the frontend.

The feature is behind a feature toggle (`pluginsSriChecks`)

**Why do we need this feature?**

Add SRI checks for plugin frontend assets.

**Who is this feature for?**

Everyone.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Part of https://github.com/grafana/grafana-community-team/issues/15 

**Special notes for your reviewer:**

Enterprise PR: ~~https://github.com/grafana/grafana-enterprise/pull/7175~~

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

xnyo committed:
Plugins: Pass hashes for SRI to frontend

xnyo committed:
Add SRI hashes to frontendsettings DTOs

xnyo committed:
Add docstring

xnyo committed:
TestSriHashes

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
Fix typo

xnyo committed:
Changed SriHashes to ModuleHash

xnyo committed:
update loader_test compareOpts

xnyo committed:
update ModuleHash error message

xnyo committed:
Add TestModuleHash/no_module.js

xnyo committed:
Add omitEmpty to moduleHash

xnyo committed:
Add ModuleHash to api/plugins/${pluginId}/settings

xnyo committed:
moved ModuleHash field

jackw committed:
feat(plugins): add moduleHash to bootData and plugin types

jackw committed:
feat(plugins): if moduleHash is available apply it to systemjs importmap

xnyo committed:
Merge branch 'giuseppe/plugins-sri-backend' of github.com:grafana/grafana into giuseppe/plugins-sri-backend

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
Calculate ModuleHash for CDN provisioned plugins

xnyo committed:
Add ModuleHash tests for TestCalculate

xnyo committed:
adjust test case name

xnyo committed:
removed .envrc

xnyo committed:
Fix signature verification failing for internal plugins

xnyo committed:
fix tests

xnyo committed:
Add pluginsFilesystemSriChecks feature togglemk

xnyo committed:
renamed FilesystemSriChecksEnabled

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

jackw committed:
refactor(plugin_loader): prefer extending type declaration over ts-error

xnyo committed:
added a couple more tests

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

marefr commented on pkg/plugins/manager/signature/manifest.go:
```diff
@@ -57,30 +59,47 @@ func (m *PluginManifest) isV2() bool {
 
+// ModuleHash returns the SRI hash for the module.js file, as expected by the browser.
+// It returns an empty string if there's no module.js in the manifest.
+func (m *PluginManifest) ModuleHash() (string, error) {
```
Could there be nested module.js files? 🤔

marefr replied:
For nested plugins I mean

jackw replied:
Good point. Yes there can be nested module.js files in the manifest:

```
"module.js": "efd8db933b56abcfd33f8c4bd66b29b074bc8e696dab6d2efff683e0cae6ca3c",
"datasource/module.js": "d5d85ee941585c34a9a3f0ceb38e45ef16c3691aae3e1f056efa1db2ce3cb251",
"panel-triggers/module.js": "7ee059fb3b8f4d28e6f5468486e22d07ab34fd0d6ce79c168227f0ac8c7a0353",
```

xnyo replied:
Great catch. Fixed with the latest iteration :smiley:

wbrowne reviewed (commented):
Do you think it would be a lot of work to move this out of the core packages and have the module hash calculated within a service inside `pluginsintegration`? I kind of struggled with the same problem with loading strategy and decided to keep it out in the end. 

It's not a deal breaker IMO, but I just had the initial feeling that maybe it's cleaner to separate out - either into a service or into the pluginstore.Plugin model instead for example. WDYT?

jackw commented:
/deploy-to-hg --enterprise-ref giuseppe/plugins-sri-backend

xnyo commented on pkg/plugins/manager/pipeline/bootstrap/steps.go:
```diff
@@ -18,12 +18,13 @@ import (
 type DefaultConstructor struct {
 	pluginFactoryFunc   pluginFactoryFunc
 	signatureCalculator plugins.SignatureCalculator
+	features            config.Features
```
Good catch, that was a leftover from a previous iteration. I'll remove it :+1:

xnyo committed:
Removed unused features

xnyo committed:
Removed unused argument from signature.DefaultCalculator call

xnyo committed:
Removed unused argument from bootstrap.DefaultConstructFunc

xnyo committed:
Moved ModuleHash to pluginassets service

xnyo committed:
update docstring

xnyo committed:
lint

xnyo committed:
Removed cdn dependency from manifest.Signature

xnyo committed:
add tests

xnyo committed:
fix extra parameters in tests

xnyo committed:
"fix" tests

xnyo committed:
removed outdated test

xnyo commented:
Moving into draft again based on the PR reviews above. I will rework so it works properly with nested plugins and refactor into pluginsintegration

xnyo commented on pkg/api/frontendsettings.go:
```diff
@@ -539,8 +541,9 @@ func (hs *HTTPServer) getFSDataSources(c *contextmodel.ReqContext, availablePlug
 					Module:    ds.Module,
-					BaseURL:   ds.BaseURL,
-					Angular:   ds.Angular,
+					// ModuleHash: hs.pluginAssets.ModuleHash(c.Req.Context(), ds),
+					BaseURL: ds.BaseURL,
+					Angular: ds.Angular,
```
This is for `ds.BuiltIn`. Do we need the ModuleHash in this case @jackw ? We also don't have the LoadingStrategy in this case, so not sure

xnyo committed:
removed unused cdn dependency in signature.DefaultCalculator

xnyo committed:
reduce diff

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
Cache returned values

xnyo committed:
Add support for deeply nested plugins (more than 1 hierarchy level)

xnyo commented on pkg/plugins/manager/signature/manifest.go:
```diff
@@ -167,6 +189,36 @@ func (s *Signature) Calculate(ctx context.Context, src plugins.PluginSource, plu
+	// Try to calculate module.js hash for SRI checks.
+	// Do not calculate the hash for filesystem plugins, unless the corresponding feature toggle is enabled.
+	var moduleHash string
+	if s.cdn.PluginSupported(plugin.JSONData.ID) || s.cfg.Features.FilesystemSriChecksEnabled {
```
I have moved everything into pluginsintegration so this dependency between signature and cdn has also been removed as a side effect

xnyo committed:
simplify cache usage

xnyo committed:
refactor TestService_ModuleHash_Cache

xnyo committed:
removed unused testdata

xnyo commented:
I have reworked the backend changes:
- Fix support for nested plugins
- Moved module hash calculation into `pluginsintegration/pluginassets`

I also have to add-back the tests for `frontendsettings`, but I think I'll leave them for a follow-up PR since moduleHash for FS plugins is behind a feature toggle for now.

PTAL :slightly_smiling_face:

xnyo commented:
/deploy-to-hg

wbrowne commented on pkg/services/pluginsintegration/pluginassets/pluginassets.go:
```diff
@@ -69,6 +82,87 @@ func (s *Service) LoadingStrategy(_ context.Context, p pluginstore.Plugin) plugi
+		return "", nil
+	}
+
+	if p.Parent != nil {
```
If you wanted to make life a bit easier for yourself here, you could turn `p.Parent`'s type into pluginstore.Plugin so we don't have to go fetch it manually.

wbrowne commented on pkg/services/pluginsintegration/pluginassets/pluginassets.go:
```diff
@@ -69,6 +82,87 @@ func (s *Service) LoadingStrategy(_ context.Context, p pluginstore.Plugin) plugi
+		return s.moduleHash(ctx, parent, childFSBase)
+	}
+
+	manifest, err := s.signature.ReadPluginManifestFromFS(ctx, p.FS)
```
Maybe a nice improvement down the road would be to add `Files map[string]string` to [the Signature struct](https://github.com/grafana/grafana/blob/main/pkg/plugins/models.go#L166) so we can reach this information more easily here.

xnyo replied:
Good point! This will also remove the need for the cache, as the MANIFEST.txt will be read only once and then the hashes will stay in memory. I'll leave it for a follow-up PR just for the sake of getting a first version up and running in Cloud

wbrowne commented on pkg/services/pluginsintegration/pluginassets/pluginassets_test.go:
```diff
@@ -179,21 +181,295 @@ func TestService_Calculate(t *testing.T) {
+	}
+}
+
+func withAngular(angular bool) func(p pluginstore.Plugin) pluginstore.Plugin {
```
Niceeee 👌

wbrowne reviewed (approved):
LGTM mate - awesome work on this!

marefr commented on pkg/services/pluginsintegration/pluginassets/pluginassets.go:
```diff
@@ -69,6 +82,87 @@ func (s *Service) LoadingStrategy(_ context.Context, p pluginstore.Plugin) plugi
+// If the plugin is unsigned, an empty string is returned.
+// The results are cached to avoid repeated reads from the MANIFEST.txt file.
+func (s *Service) ModuleHash(ctx context.Context, p pluginstore.Plugin) string {
+	cachedValue, ok := s.moduleHashCache.Load(p.ID)
```
I guess there's no case where a new plugin is installed during runtime (since cloud only)? Otherwise might need to include version in cache key.

xnyo replied:
Good spot, this can also work on-prem (although behing a feature flag), so I added the version to the key in https://github.com/grafana/grafana/pull/93024/commits/44e2e81b2d7b93293e394e3389bc322f94ef1915

marefr commented on pkg/services/pluginsintegration/pluginstore/plugins.go:
```diff
@@ -10,7 +10,7 @@ import (
 	plugins.JSONData
 
-	fs                plugins.FS
+	FS                plugins.FS
```
Nit. Feels like a leaky abstraction for a DTO, but understand maybe tricky to workaround/find a better solution? Maybe something to follow up on...

marefr reviewed (approved):
LGTM with some questions/nits

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
re-generate feature toggles

xnyo committed:
use version for module hash cache

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
Renamed feature toggle to pluginsSriChecks and use it for both cdn and filesystem

xnyo committed:
Removed app/types/system-integrity.d.ts

xnyo committed:
re-generate feature toggles

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend

xnyo committed:
re-generate feature toggles

academo commented on public/app/features/plugins/plugin_loader.ts:
```diff
@@ -94,7 +96,17 @@ export async function importPluginModule({
+  const resolvedModule = System.resolve(modulePath);
+  const integrityMap = System.getImportMap().integrity;
+
+  if (moduleHash && integrityMap && !integrityMap[resolvedModule]) {
```
since all this functionality is behind a feature flag, we need to also make sure the frontend code only executes is the feature flag is on to prevent any possible side effects

mckn replied:
Sounds like a good suggestion by @academo 👍🏻

jackw replied:
Thanks for the suggestion. I've updated the logic in this commit: https://github.com/grafana/grafana/pull/93024/commits/4c9a1cf6ac387437ff6673d4da2ec86a13c79c97

mckn reviewed (approved):
LGTM! It is probably a good idea to add a check to see if the feature flag i enabled in the frontend as well.

jackw committed:
feat(plugins): put systemjs integrity hash behind feature flag

xnyo committed:
Merge branch 'main' into giuseppe/plugins-sri-backend
